dialogue-core/src/main/java/com/palantir/dialogue/core/ChannelToEndpointChannel.java

```diff
@@ -16,54 +16,25 @@
 
 package com.palantir.dialogue.core;
 
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.palantir.dialogue.Channel;
 import com.palantir.dialogue.Endpoint;
 import com.palantir.dialogue.Request;
 import com.palantir.dialogue.Response;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Function;
 
 final class ChannelToEndpointChannel implements Channel {
 
-    private final Function<Endpoint, Channel> adapter;
-    private final Map<Object, Channel> cache;
+    private final LoadingCache<Endpoint, Channel> cache;
 
-    ChannelToEndpointChannel(Function<Endpoint, Channel> adapter) {
-        this.adapter = adapter;
-        this.cache = new ConcurrentHashMap<>();
+    ChannelToEndpointChannel(Function<Endpoint, Channel> loader) {
+        this.cache = Caffeine.newBuilder().weakKeys().maximumSize(10_000).build(loader::apply);
```

Comment:
I suppose if we were willing to give up the maximum size we could just use a `ConcurrentHashMap` here and avoid some of the `Caffeine` bits. Although maybe it's worth it to have the maximum size.

Reply:
Also weak keys, which are key to making this somewhat safe

Reply:
Ah right, we'd still have to periodically clean up the entries for collected keys.